STM32/FLASH_readwrite/Src/FLASH_readwrite.c

```diff
@@ -2,139 +2,158 @@
 /*
  *  Library:			Read and write FLASH memory on STM32F411
  *  Description:		Provides an interface to read and write FLASH memory.
- *  TODO: This is a very brute force implementation where no 'protocol' is used
- *        to read/write the flash. So flash layout should be defined.
 */
 #if defined(STM32F401xC)
-
-#include <stm32f4xx_hal.h>
 #include <FLASH_readwrite.h>
 #include <string.h>
 
-// NOTE: The flash sector and addr variables originate from the ld linker script.
-// The flash is used for program data and this must NOT be cleared. Putting data in
-// linker script will ensure that program data will not be cleared when writing
-// calibration data to the flash. To access these variables the syntax below is used.
-extern uint32_t _FlashSector; // Variable defined in ld linker script.
-extern uint32_t _FlashAddr;   // Variable defined in ld linker script.
-#define FLASH_SECTOR ((uint32_t) &_FlashSector)
-#define FLASH_ADDR   ((uint32_t) &_FlashAddr)
-
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-// Write data to FLASH_ADDR+indx directly. This method does not
-// ensure any data integrity as data validation is performed.
-void writeToFlash(uint32_t indx, uint32_t size, uint8_t *data)
+static uint32_t getFlashSector(uint32_t address)
 {
-    __HAL_RCC_WWDG_CLK_DISABLE();
+    uint32_t address_offset = address - START_SECTOR_0;
+    uint32_t address_sectors_start[6] = {START_SECTOR_0, START_SECTOR_1, START_SECTOR_2, START_SECTOR_3,
+                                         START_SECTOR_4, START_SECTOR_5};
+
+    uint32_t address_sectors_end[6] = {END_SECTOR_0, END_SECTOR_1, END_SECTOR_2, END_SECTOR_3,
+                                       END_SECTOR_4, END_SECTOR_5};                                         
+    for (int i = 0; i < 6; i++)
+    {
+        if (address_offset >= address_sectors_start[i]-START_SECTOR_0 &&
+            address_offset <= address_sectors_end[i]-START_SECTOR_0)
```

Comment:
This seems slightly odd:
First you create the address offset by subtracting START_SECTOR_0
Then you compare it to all the address sectors which you also subtract START_SECTOR_0 from.

Just get rid of all the subtracting? Or have I missed something

Reply:
If it seems odd, it is because it is :smile:  This was because the getFlashSector() was inspired by the getSector() in the H7 implementation where you have sectors 0-7 in two memory banks. 

It is not needed and has been removed.